### ISSUE TITLE
View Badges Security Bug

### ISSUE THREAD

**Issue by radams22** [SparkDevNetwork/Rock#4684]
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Did you perform a search at https://github.com/issues?q=is%3Aissue+user%3ASparkDevNetwork+-repo%3ARock to see if your bug or enhancement is already reported?
    * Can you reproduce the problem on a fresh install or the [demo site](http://rock.rocksolidchurchdemo.com/)?
    * Did you include your Rock version number and [client culture](https://github.com/SparkDevNetwork/Rock/wiki/Environment-and-Diagnostics-Information) setting?

### Description

Unable to view certain badges based on security role. 

### Steps to Reproduce

1. Create a new security role and add a group member of someone who DOES NOT EXIST in the following security roles:
      1. RSR - Rock Administration
      2. RSR - Staff Workers
      3. RSR - Staff Like Workers
2. Add a new security role to each block on the person profile page with the security permissions of View.
3. Add a new security role to the person profile page with security permissions of View.
4. Enable Impersonation.
5. Impersonate the Group Member who exists in the new security role.

**Expected behavior:**

The group member in the new security role should see the badges they've been given explicit permissions to view.

**Actual behavior:**

The group member does not see any badges of the following types:
1. In Group of Type
2. Family Attendance
3. Family Weeks Attended in Duration
4. Steps

### Versions
[Badge Security Bug.docx](https://github.com/SparkDevNetwork/Rock/files/6473529/Badge.Security.Bug.docx)

* **Rock Version:** [V12.2,12.3,13.0]
* **Client Culture Setting:** [US EN]

**Comment by JimMichael** [comment]
I don't believe this is a bug with badges themselves. We have a non-staff/admin role that has access to the Person Profile page as well, and they can see all badges the role has access to just fine... the difference (likely) being that we granted this role view access to the _site_ itself. Taking that right away still lets the role get to the profile page, but badges are gone.

Early on in our process of giving a non-staff/admin role access to internal Rock, we realized that there was SO MUCH STUFF wired up that only works when you have view rights to the *site*, that it was easier to just grant that to the new role, then *deny* that role anywhere we didn't want it to have access (mostly top-level menus).

**Comment by radams22** [comment]
@JimMichael  Thanks Jim, we actually tried to give them the site access, but they are still unable to see the badges of a specific type. Some they can see, others they can't. So I agree, it's not a badge issue, it's a security issue that we cannot figure out.

**Comment by cwomack** [comment]
@radams22 Let me illustrate the best way to ensure that the badges are only visible to specific users across the entirety of Rock.  Rather than trying to set up security on every block or certain pages manually, you'll want to set the security permissions on the **badge itself** with the new role you've made.  

Can you try to setting security on the badge itself with the following steps and see if it fixes the problem you're having?

1. Create a security role as you had already described above under **Admin Tools > Security > Security Roles**.  I'll call mine Badge Security for the following screenshots.

![chrome_nJXmnL6m51](https://user-images.githubusercontent.com/67726635/118541438-84c32380-b706-11eb-8d5b-5b385ad1ac5b.png)

2. Go under **Admin Tools > General Settings > Badges** and with the badge you'd like to secure, click on the "lock" on the right hand side of the list.

![CWJ9Cqig9I](https://user-images.githubusercontent.com/67726635/118540579-91934780-b705-11eb-8c2a-87fc5e15999e.png)

3. Add the newly made role and ensure it's _ABOVE_ the "Deny All Users."  **Order matters here**!

![chrome_eXiWV4lgoR](https://user-images.githubusercontent.com/67726635/118542206-8c36fc80-b707-11eb-8677-89676234fcf8.png)

4. Log in (or impersonate) with people who do/do not have role to ensure it's working as you'd expect. **Note - I had to give Jim Simmons the "Staff Like Worker" role as well so he could see the person profile page**.

![chrome_IL2kOxSHG4](https://user-images.githubusercontent.com/67726635/118543339-0320c500-b709-11eb-9e88-fd71089b482e.png)

**Comment by radams22** [comment]
@cwomack I noticed in your screenshot that you have "All Users" set at Deny for the badge security. All of our badges are set to allow "All Users" view access. I double-checked the block setting on the profile as well and ensured "All Users" had view access to it as well, and they do. 

I did some additional testing and it seems if the user does not have at least the Staff Like Worker security role, there are certain badge types they are unable to see, which is listed in my original outline above. Even though the badge security is set to view access for all users. 

The reason we do not use Staff Like Worker for all of the users helping within Rock is that role gives too much access for our volunteers helping in this area of need. 

I have attached some more screen shots of what our security is set to for the badges and the block. As well as showing what they can see with and without staff-like access.

[Badge View Access Examples.docx](https://github.com/SparkDevNetwork/Rock/files/6502384/Badge.View.Access.Examples.docx)

**Comment by cwomack** [comment]
@radams22 We are looking into this further and have reproduced what you've reported.  We will update this issue once we have more information.  Thank you for your patience and replies.

**Comment by ethan-sparkdevnetwork** [comment]
@radams22 I located the issue. The badges that are not appearing are using REST controllers to get the information. The security settings for the REST controllers include Staff Like Workers (and up) but deny View permission to "All Users". You can get this to work right away by adding the roll "All Authenticated Users" to Allow View.

Admin Tools --> Security --> REST Controllers. Scroll down to "Badges". You can set the security for every badge controller by clicking the lock icon. If you want to be more selective then click the row and you will get a list of all the controllers for badges which you can set using the lock icon.

We will review this internally to see if the default security settings for this are too restrictive.
